From 6de76a04bad7d6147c271c0b7eeeee5d322332c3 Mon Sep 17 00:00:00 2001
From: Abinaya Dhandapani <adhandap@amd.com>
Date: Thu, 14 Mar 2024 01:52:53 -0500
Subject: [PATCH] com.amd: AMD OEM RAS configuration interface added

This commit introduces an AMD OEM interface that enables the setting
and retrieval of configuration parameters used by the RAS application.
For more details on the AMD RAS feature, please refer to the design
document uploaded for review.
https://gerrit.openbmc.org/c/openbmc/docs/+/68440

Change-Id: I51d4addfbc2231a3dcd5533033a693f882f0200f
Signed-off-by: Abinaya Dhandapani <adhandap@amd.com>
---
 gen/com/amd/RAS/Configuration/meson.build     | 38 ++++++++++
 gen/com/amd/RAS/meson.build                   | 27 +++++++
 gen/com/amd/crashdump/meson.build             | 38 ++++++++++
 gen/com/amd/meson.build                       | 28 ++++++++
 gen/com/meson.build                           |  1 +
 meson.build                                   |  3 +
 meson.options                                 |  1 +
 yaml/com/amd/RAS/Configuration.interface.yaml | 72 +++++++++++++++++++
 yaml/com/amd/crashdump.interface.yaml         | 22 ++++++
 9 files changed, 230 insertions(+)
 create mode 100644 gen/com/amd/RAS/Configuration/meson.build
 create mode 100644 gen/com/amd/RAS/meson.build
 create mode 100644 gen/com/amd/crashdump/meson.build
 create mode 100644 gen/com/amd/meson.build
 create mode 100644 yaml/com/amd/RAS/Configuration.interface.yaml
 create mode 100644 yaml/com/amd/crashdump.interface.yaml

diff --git a/gen/com/amd/RAS/Configuration/meson.build b/gen/com/amd/RAS/Configuration/meson.build
new file mode 100644
index 0000000..27e502a
--- /dev/null
+++ b/gen/com/amd/RAS/Configuration/meson.build
@@ -0,0 +1,38 @@
+# Generated file; do not modify.
+
+sdbusplus_current_path = 'com/amd/RAS/Configuration'
+
+generated_sources += custom_target(
+    'com/amd/RAS/Configuration__cpp'.underscorify(),
+    input: ['../../../../../yaml/com/amd/RAS/Configuration.interface.yaml'],
+    output: [
+        'common.hpp',
+        'server.hpp',
+        'server.cpp',
+        'aserver.hpp',
+        'client.hpp',
+    ],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'cpp',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../../../yaml',
+        'com/amd/RAS/Configuration',
+    ],
+    install: should_generate_cpp,
+    install_dir: [
+        get_option('includedir') / sdbusplus_current_path,
+        get_option('includedir') / sdbusplus_current_path,
+        false,
+        get_option('includedir') / sdbusplus_current_path,
+        get_option('includedir') / sdbusplus_current_path,
+    ],
+    build_by_default: should_generate_cpp,
+)
+
diff --git a/gen/com/amd/RAS/meson.build b/gen/com/amd/RAS/meson.build
new file mode 100644
index 0000000..2068756
--- /dev/null
+++ b/gen/com/amd/RAS/meson.build
@@ -0,0 +1,27 @@
+# Generated file; do not modify.
+subdir('Configuration')
+
+sdbusplus_current_path = 'com/amd/RAS'
+
+generated_markdown += custom_target(
+    'com/amd/RAS/Configuration__markdown'.underscorify(),
+    input: ['../../../../yaml/com/amd/RAS/Configuration.interface.yaml'],
+    output: ['Configuration.md'],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'markdown',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../../yaml',
+        'com/amd/RAS/Configuration',
+    ],
+    install: should_generate_markdown,
+    install_dir: [inst_markdown_dir / sdbusplus_current_path],
+    build_by_default: should_generate_markdown,
+)
+
diff --git a/gen/com/amd/crashdump/meson.build b/gen/com/amd/crashdump/meson.build
new file mode 100644
index 0000000..b2ead0a
--- /dev/null
+++ b/gen/com/amd/crashdump/meson.build
@@ -0,0 +1,38 @@
+# Generated file; do not modify.
+
+sdbusplus_current_path = 'com/amd/crashdump'
+
+generated_sources += custom_target(
+    'com/amd/crashdump__cpp'.underscorify(),
+    input: ['../../../../yaml/com/amd/crashdump.interface.yaml'],
+    output: [
+        'common.hpp',
+        'server.hpp',
+        'server.cpp',
+        'aserver.hpp',
+        'client.hpp',
+    ],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'cpp',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../../yaml',
+        'com/amd/crashdump',
+    ],
+    install: should_generate_cpp,
+    install_dir: [
+        get_option('includedir') / sdbusplus_current_path,
+        get_option('includedir') / sdbusplus_current_path,
+        false,
+        get_option('includedir') / sdbusplus_current_path,
+        get_option('includedir') / sdbusplus_current_path,
+    ],
+    build_by_default: should_generate_cpp,
+)
+
diff --git a/gen/com/amd/meson.build b/gen/com/amd/meson.build
new file mode 100644
index 0000000..0b2ef8b
--- /dev/null
+++ b/gen/com/amd/meson.build
@@ -0,0 +1,28 @@
+# Generated file; do not modify.
+subdir('RAS')
+subdir('crashdump')
+
+sdbusplus_current_path = 'com/amd'
+
+generated_markdown += custom_target(
+    'com/amd/crashdump__markdown'.underscorify(),
+    input: ['../../../yaml/com/amd/crashdump.interface.yaml'],
+    output: ['crashdump.md'],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'markdown',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../yaml',
+        'com/amd/crashdump',
+    ],
+    install: should_generate_markdown,
+    install_dir: [inst_markdown_dir / sdbusplus_current_path],
+    build_by_default: should_generate_markdown,
+)
+
diff --git a/gen/com/meson.build b/gen/com/meson.build
index a0f9a5c..8aebec5 100644
--- a/gen/com/meson.build
+++ b/gen/com/meson.build
@@ -1,4 +1,5 @@
 # Generated file; do not modify.
+subdir('amd')
 subdir('ampere')
 subdir('google')
 subdir('ibm')
diff --git a/meson.build b/meson.build
index ba7b7db..40645e5 100644
--- a/meson.build
+++ b/meson.build
@@ -27,6 +27,9 @@ yaml_selected_subdirs = []
 if get_option('data_com_ampere')
     yaml_selected_subdirs += 'com/ampere'
 endif
+if get_option('data_com_amd')
+    yaml_selected_subdirs += 'com/amd'
+endif
 if get_option('data_com_google')
     yaml_selected_subdirs += 'com/google'
 endif
diff --git a/meson.options b/meson.options
index 643f67e..4982ea1 100644
--- a/meson.options
+++ b/meson.options
@@ -1,6 +1,7 @@
 option('generate_md', type: 'boolean', value: true)
 option('libphosphor_dbus', type: 'boolean', value: true)
 option('data_com_ampere', type: 'boolean', value: true)
+option('data_com_amd', type: 'boolean', value: true)
 option('data_com_google', type: 'boolean', value: true)
 option('data_com_ibm', type: 'boolean', value: true)
 option('data_com_intel', type: 'boolean', value: true)
diff --git a/yaml/com/amd/RAS/Configuration.interface.yaml b/yaml/com/amd/RAS/Configuration.interface.yaml
new file mode 100644
index 0000000..bedd92d
--- /dev/null
+++ b/yaml/com/amd/RAS/Configuration.interface.yaml
@@ -0,0 +1,72 @@
+description: >
+    Configuration parameters used by the AMD BMC RAS application.
+
+methods:
+    - name: SetAttribute
+      description: >
+          To set the new value on existing attribute name.
+      parameters:
+          - name: AttributeName
+            type: string
+            description: >
+                AttributeName which has to be changed.
+          - name: AttributeValue
+            type:
+                variant[boolean,string,int64,array[string],dict[string, string]]
+            description: >
+                New attribute value
+
+      errors:
+          - xyz.openbmc_project.Common.Error.InvalidArgument
+
+    - name: GetAttribute
+      description: >
+          To get the attribute value of existing attributes.
+      parameters:
+          - name: AttributeName
+            type: string
+            description: >
+                To get the RAS configuration attribute current values.
+
+      returns:
+          - name: Value
+            type:
+                variant[boolean,string,int64,array[string],dict[string, string]]
+            description: >
+                The attribute value.
+
+      errors:
+          - xyz.openbmc_project.Common.Error.InvalidArgument
+          - xyz.openbmc_project.Common.Error.InternalFailure
+          - xyz.openbmc_project.Common.File.Error.Open
+
+properties:
+    - name: RasConfigTable
+      type:
+          dict[string,struct[enum[self.AttributeType],string,variant[boolean,string,int64,array[string],dict[string,
+          string]],int64]]
+      description: >
+          map{attributeName,struct{attributeType,description,value,maxboundlimit}
+          Example : {"McaErrCounterOverflow",
+          {com.amd.RAS.Configuration.AttributeType.Integer,
+           "Error counter threshold overflow for MCA errors",10,0xFFF}
+enumerations:
+    - name: AttributeType
+      description: >
+          Attribute Type.
+      values:
+          - name: Boolean
+            description: >
+                Boolean value Type
+          - name: String
+            description: >
+                string value Type.
+          - name: Integer
+            description: >
+                unsigned integer value Type.
+          - name: ArrayOfStrings
+            description: >
+                Array of strings value Type.
+          - name: KeyValueMap
+            description: >
+                key-value pairs value Type.
diff --git a/yaml/com/amd/crashdump.interface.yaml b/yaml/com/amd/crashdump.interface.yaml
new file mode 100644
index 0000000..a0035d8
--- /dev/null
+++ b/yaml/com/amd/crashdump.interface.yaml
@@ -0,0 +1,22 @@
+description: >
+    This interface provides details about Common Platform Error Record (CPER)
+    files that are available on the BMC. Each CPER file is represented as a CPER
+    Record object, which holds key attributes of the file.
+properties:
+    - name: Filename
+      type: string
+      description: >
+          The name of the CPER file stored on the BMC. This attribute uniquely
+          identifies the file and can be used to locate or refer to the specific
+          CPER file in the system.
+    - name: Timestamp
+      type: string
+      description: >
+          Indicates the time when the CPER file was generated. This timestamp
+          provides critical information for correlating the error record with
+          system events or logs, aiding in diagnostics and root cause analysis.
+    - name: Log
+      type: string
+      description: >
+          An additional string that gives the file path of the CPER files
+          saved in the BMC.
-- 
2.52.0

